### PR TITLE
add some redis client config that may be commonly used

### PR DESCRIPTION
**Pull request by Taction**
# Description

add some config that can be set from metadata for redis client.

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [x] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#1334

RELEASE NOTES: Added commonly used Redis client settings

### CODE REVIEW

File: pubsub/redis/metadata.go
Comment (pkedy) [inline pubsub/redis/metadata.go:18]: For consistency, this should be parsed as a `time.Duration` like `redeliverInterval` and `processingTimeout` are and the `InSec` suffix removed.

File: pubsub/redis/metadata.go
Comment (pkedy) [inline pubsub/redis/metadata.go:21]: Same

File: pubsub/redis/metadata.go
Comment (pkedy) [inline pubsub/redis/metadata.go:23]: Same

File: pubsub/redis/metadata.go
Comment (pkedy) [inline pubsub/redis/metadata.go:27]: Same

File: pubsub/redis/redis.go
Comment (pkedy) [inline pubsub/redis/redis.go:137]: Note how `redeliverInterval` and `processingTimeout` are parsed. First as uint milliseconds (e.g. `5000`), then as a duration (e.g. `5s`).

**Review by pkedy — CHANGES_REQUESTED** [review]
I love seeing more connection settings coming in this PR! I'd like just a few changes for consistency with the duration-based settings.

File: pubsub/redis/metadata.go
Comment (Taction) [reply pubsub/redis/metadata.go:18]: I will add more  connection settings, and keep their format identical to definition in redis client config as much as possible.

File: pubsub/redis/redis.go
Comment (Taction) [reply pubsub/redis/redis.go:137]: Thanks for advice, I will Parse duration-based settings like how redeliverInterval and processingTimeout are parsed.

File: pubsub/redis/redis.go
Comment (pkedy) [inline pubsub/redis/redis.go:26]: My apologies, I didn't catch this before. In other components this field is called `backoffMaxRetries`, mainly because the [github.com/cenkalti/backoff](https://github.com/cenkalti/backoff) library is used.
Please rename this to `backoffMaxRetries`. It's just a consistency thing across components even though the Redis setting is differently named.

File: pubsub/redis/redis.go
Comment (pkedy) [inline pubsub/redis/redis.go:27]: This one would be `backOffMinInterval`.

File: pubsub/redis/redis.go
Comment (pkedy) [inline pubsub/redis/redis.go:28]: This one would be `backOffMaxInterval`.

**Review by pkedy — CHANGES_REQUESTED** [review]

**Comment by pkedy** [comment]
@Taction Thanks for all your work on this. I know there have been a few iterations now. Two things:

1) Before this can be merged, there should be an open PR in dapr/docs to update [this page](https://docs.dapr.io/operations/components/setup-pubsub/supported-pubsub/setup-redis-pubsub/) with the new settings. **Make sure branch from and target the `v1.1` branch in the PR. Its easy to forget.**

2) I think I have steered you in the wrong direction with the settings naming but wanted to ask your opinion.

*Context*: In `v1.2` we'll likely be adding more backoff settings to some of the pubsub components. The backoff/retry library we are using handles retries between the sidecar and the application.

*Where I steered you wrong*: All the settings I had you prefix with `backOff` are actually between the sidecar and Redis itself. Correct? If that is the case a different prefix should be used. Maybe `read`? I'm not actually sure where exactly these settings come into play in the client module but I'm assuming its for any CMD call. WDYT?

**Comment by Taction** [comment]
> @Taction Thanks for all your work on this. I know there have been a few iterations now. Two things:
> 
> 1. Before this can be merged, there should be an open PR in dapr/docs to update [this page](https://docs.dapr.io/operations/components/setup-pubsub/supported-pubsub/setup-redis-pubsub/) with the new settings. **Make sure branch from and target the `v1.1` branch in the PR. Its easy to forget.**
> 2. I think I have steered you in the wrong direction with the settings naming but wanted to ask your opinion.
> 
> _Context_: In `v1.2` we'll likely be adding more backoff settings to some of the pubsub components. The backoff/retry library we are using handles retries between the sidecar and the application.
> 
> _Where I steered you wrong_: All the settings I had you prefix with `backOff` are actually between the sidecar and Redis itself. Correct? If that is the case a different prefix should be used. Maybe `read`? I'm not actually sure where exactly these settings come into play in the client module but I'm assuming its for any CMD call. WDYT?

1. For the second thing, I know what you actually want to tell me. I've seen the use of backoff options in mqtt component which is used for handle retries between the sidecar and the application.But there are lots components do not have these options so  I didn't catch the point at that time. And I should have confirmed with you about it.

As for the prefix, is `redis` a better choice（such as `redisMaxRetries`）? As far as I'm concerned it may have 2 potential advantage. One is that it may be more easily for users to tell these options apart from `backOff` options, as the name may shows that it is for redis configuration. Two is that `read` is a more general expression, if some other components have similar option between client and server `read` may be a better choice(I am not very familiar with other components now, so i assume that there is not much components whit these option).

2. For the first thing,I will work on this but it will take me some time to finish it. And It is very kind of you to stress out what is easy to forget.

What's more, in my opinion both `read` and `redis` is fine for the prefix. Which one do you think is more suitable?

**Comment by pkedy** [comment]
@Taction I think we are in agreement! `redis` or `read` is fine as a prefix. As the author of this PR, I'll let you decide between the two. :) I just wanted to avoid `backOff` in this case because its for a different point in the call chain. Again-my mistake for steering you that way to begin with and thanks for your understanding. Once you make that change, we can finally get this merged unless @artursouza has other comments.

**Comment by artursouza** [comment]
> @Taction
> 
> For 1, the `backOff*` settings are slowly being added to the components starting with the ones that really really need it (e.g. Kafka). Also, `backOffMaxRetries` is only setting for now but [more settings are one their way maybe in 1.2](https://github.com/dapr/components-contrib/compare/master...pkedy:using_config_package_part_1).
> 
> For 2, I think we are in agreement! `redis` or `read` is fine as a prefix. As the author of this PR, I'll let you decide between the two. :) I just wanted to avoid `backOff` in this case because its for a different point in the call chain. Again-my mistake for steering you that way to begin with and thanks for your understanding. Once you make that change, we can finally get this merged unless @artursouza has other comments.
> 
> Also, please mention the docs PR in this PR or visa versa. Thanks!


LGTM

**Comment by pkedy** [comment]
@Taction To @mukundansundar 's point in your docs PR, we need an issue created as this is a pretty good sized enhancement. We do this for tracking things int the RELEASE NOTES.

**Comment by pkedy** [comment]
@Taction I added RELEASE NOTES to the description of this PR which our release process also picks up. So in the interest of moving this along, no need to create an issue. Just know for the future that large changes need an issue.

**Review by pkedy — APPROVED** [review]

**Pull request merged**
